charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,249 @@
+        self._log_and_set_status(MaintenanceStatus("Removing K8S resources"))
+        try:
+            self.k8s_resource_handler.delete()
+        except ApiError as error:
+            # Do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Removing K8S resources failed with error: {error}")
+                raise GenericCharmRuntimeError("Removing K8s resources failed") from error
+
+    def _on_event(self, event) -> None:
+        """Perform required actions for every event."""
+        try:
+            self._check_leader()
+            self._deploy_k8s_resources()
+            update_layer(
+                self._container_name,
+                self._container,
+                self._tensorboards_web_app_layer,
+                self.logger,
+            )
+            interfaces = self._get_interfaces()
+            self._configure_mesh(interfaces)
+            self.unit.status = ActiveStatus()
+
+        except ErrorWithStatus as error:
+            self._log_and_set_status(error.status)
+            return
+
+        self.unit.status = ActiveStatus()
+
+    def _deploy_k8s_resources(self) -> None:
+        """Deploy K8S resources."""
+        try:
+            self._log_and_set_status(MaintenanceStatus("Creating K8S resources"))
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            raise ErrorWithStatus("Creating K8S resources failed", BlockedStatus)
+        self._log_and_set_status(MaintenanceStatus("K8S resources created"))
+
+    def _configure_mesh(self, interfaces) -> None:
         if interfaces["ingress"]:
             interfaces["ingress"].send_data(
                 {
                     "prefix": "/tensorboards",
                     "rewrite": "/",
-                    "service": self.model.app.name,
-                    "port": self.model.config["port"],
+                    "service": self.app.name,
+                    "port": PORT,
                 }
             )
+        else:
+            raise ErrorWithStatus("No ingress relation available", BlockedStatus)
 
-    def _check_leader(self):
+    def _check_leader(self) -> None:
+        """Check if this unit is a leader."""
         if not self.unit.is_leader():
             # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailed("Waiting for leadership", WaitingStatus)
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
 
     def _get_interfaces(self):
         try:
             interfaces = get_interfaces(self)
-        except NoVersionsListed as err:
-            raise CheckFailed(err, WaitingStatus)
-        except NoCompatibleVersions as err:
-            raise CheckFailed(err, BlockedStatus)
+        except NoVersionsListed as error:
+            raise ErrorWithStatus(error, WaitingStatus)
+        except NoCompatibleVersions as error:
+            raise ErrorWithStatus(error, BlockedStatus)
```

Comment:
Is this BlockedStatus something a human operator can change? If so, I'd add some actions in the error message to unblock the charm, otherwise I'd change the status to something more appropriate like Waiting or Maintenance (depends on what this is doing).

Reply:
This is one of those awkward ones.  NoCompatibleVersions means we've related two apps that don't share a schema version (eg: TWA supports schema v1, pilot supports v2).  So I guess it is Blocked because the user has related two incompatible charms?

fwiw, we raise Blocked for this in any (or almost any) other charm that makes this check, so raising it here is at least consistent.

Reply:
@ca-scribner yeah, consistency-wise sounds good and it makes sense to go to BlockedStatus on that scenario. I would ask the team to revisit the messaging, though, just to make sure users know what to do in these cases.